File: python/src/multifrost/connection.py
"""
FILE: python/src/multifrost/connection.py
PURPOSE: Expose the caller-side v5 configuration object and async handle surface.
OWNS: connect(), Connection, Handle, target validation, and caller-side remote call/query flow.
EXPORTS: ConnectOptions, connect, Connection, Handle.
DOCS: docs/spec.md; agent_chat/python_v5_api_surface_2026-03-25.md
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any, TypedDict
from uuid import uuid4

from typing_extensions import Unpack

from .errors import (
    ErrorOrigin,
    RemoteCallError,
    RouterError,
    TransportError,
    error_from_body,
)
from .frame import FrameParts, decode_body, encode_body
from .protocol import (
    ROUTER_PEER_ID,
    CallBody,
    ErrorBody,
    PeerClass,
    QueryBody,
    QueryExistsResponseBody,
    QueryGetResponseBody,
    ResponseBody,
    build_call_envelope,
    build_query_envelope,
    response,
)
from .protocol import (
    error as error_kind,
)
from .router_bootstrap import RouterBootstrapConfig, bootstrap_router
from .transport import WebSocketTransport


class ConnectOptions(TypedDict, total=False):
    caller_peer_id: str
    router_port: int
    bootstrap_timeout: float
    request_timeout: float
    register_timeout: float
    eager_validate_target: bool


def connect(target_peer_id: str, **options: Unpack[ConnectOptions]) -> Connection:
    return Connection(target_peer_id, **options)


class Connection:
    """Caller configuration only; no live socket or process state."""

    def __init__(self, target_peer_id: str, /, **options: Any) -> None:
        target_peer_id = str(target_peer_id).strip()
        if not target_peer_id:
            raise ValueError("target_peer_id must not be empty")

        self._target_peer_id = target_peer_id
        self._caller_peer_id = _clean_optional_str(options.get("caller_peer_id"))
        self._router_port = _clean_optional_int(options.get("router_port"))
        self._bootstrap_timeout = _clean_optional_float(
            options.get("bootstrap_timeout"), default=10.0
        )
        self._request_timeout = _clean_optional_float(
            options.get("request_timeout"), default=30.0
        )
        self._register_timeout = _clean_optional_float(
            options.get("register_timeout"), default=10.0
        )
        self._eager_validate_target = bool(options.get("eager_validate_target", False))

    @property
    def target_peer_id(self) -> str:
        return self._target_peer_id

    @property
    def caller_peer_id(self) -> str | None:
        return self._caller_peer_id

    @property
    def router_port(self) -> int | None:
        return self._router_port

    @property
    def bootstrap_timeout(self) -> float:
        return self._bootstrap_timeout

    @property
    def request_timeout(self) -> float:
        return self._request_timeout

    @property
    def register_timeout(self) -> float:
        return self._register_timeout

    @property
    def eager_validate_target(self) -> bool:
        return self._eager_validate_target

    def handle(self) -> Handle:
        return Handle(self)

    def handle_sync(self) -> HandleSync:
        from .sync import HandleSync

        return HandleSync(self)


class _AsyncCallNamespace:
    def __init__(self, handle: Handle) -> None:
        self._handle = handle

    def __getattr__(self, function_name: str) -> Callable[..., Awaitable[Any]]:
        async def invoke(*args: Any, namespace: str | None = None) -> Any:
            return await self._handle._invoke_remote(
                function_name,
                list(args),
                namespace=namespace,
            )

        return invoke


class Handle:
    """Async caller runtime that owns one live transport."""

    def __init__(self, connection: Connection) -> None:
        self._connection = connection
        self._caller_peer_id = connection.caller_peer_id or str(uuid4())
        self._transport: WebSocketTransport | None = None
        self.call = _AsyncCallNamespace(self)

    @property
    def peer_id(self) -> str:
        return self._caller_peer_id

    @property
    def target_peer_id(self) -> str:
        return self._connection.target_peer_id

    async def start(self) -> None:
        if self._transport is not None and self._transport.active:
            return

        bootstrap_config = RouterBootstrapConfig.from_env(self._connection.router_port)
        bootstrap_config.readiness_timeout = self._connection.bootstrap_timeout
        await bootstrap_router(bootstrap_config)

        self._transport = WebSocketTransport(
            peer_id=self._caller_peer_id,
            peer_class=PeerClass.CALLER,
            router_endpoint=bootstrap_config.endpoint,
            register_timeout=self._connection.register_timeout,
            request_timeout=self._connection.request_timeout,
        )

        try:
            await self._transport.start()
            if self._connection.eager_validate_target:
                target = await self.query_peer_get(self._connection.target_peer_id)
                self._validate_target_peer(target)
        except Exception:
            await self.stop()
            raise

    async def stop(self) -> None:
        if self._transport is None:
            return
        try:
            await self._transport.close(graceful=True)
        finally:
            self._transport = None

    async def query_peer_exists(self, peer_id: str) -> bool:
        body = QueryBody.peer_exists(peer_id)
        frame = await self._router_request(body)
        response_body = self._decode_query_exists(frame)
        return response_body.exists

    async def query_peer_get(self, peer_id: str) -> QueryGetResponseBody:
        body = QueryBody.peer_get(peer_id)
        frame = await self._router_request(body)
        return self._decode_query_get(frame)

    async def _router_request(self, body: QueryBody) -> FrameParts:
        transport = self._require_transport()
        envelope = build_query_envelope(self._caller_peer_id)
        frame = await transport.request(envelope, encode_body(body))
        if frame.envelope.kind == response:
            return frame
        if frame.envelope.kind == error_kind:
            raise self._error_from_frame(frame)
        raise TransportError(f"unexpected router reply kind: {frame.envelope.kind}")

    def _decode_query_exists(self, frame: FrameParts) -> QueryExistsResponseBody:
        payload = decode_body(frame.body_bytes)
        if not isinstance(payload, dict):
            raise TransportError(f"invalid query response body type: {type(payload)!r}")
        return QueryExistsResponseBody.from_mapping(payload)

    def _decode_query_get(self, frame: FrameParts) -> QueryGetResponseBody:
        payload = decode_body(frame.body_bytes)
        if not isinstance(payload, dict):
            raise TransportError(f"invalid query response body type: {type(payload)!r}")
        return QueryGetResponseBody.from_mapping(payload)

    async def _invoke_remote(
        self,
        function_name: str,
        args: list[Any],
        *,
        namespace: str | None = None,
    ) -> Any:
        transport = self._require_transport()
        call_body = CallBody(function=function_name, args=args, namespace=namespace)
        envelope = build_call_envelope(self._caller_peer_id, self._connection.target_peer_id)
        frame = await transport.request(envelope, encode_body(call_body))
        if frame.envelope.kind == response:
            payload = decode_body(frame.body_bytes)
            if not isinstance(payload, dict):
                raise TransportError(
                    f"invalid response body type: {type(payload)!r}"
                )
            response_body = ResponseBody.from_mapping(payload)
            return response_body.result
        if frame.envelope.kind == error_kind:
            raise self._error_from_frame(frame)
        raise TransportError(f"unexpected router reply kind: {frame.envelope.kind}")

    def _error_from_frame(self, frame: FrameParts) -> RouterError | RemoteCallError:
        payload = decode_body(frame.body_bytes)
        if not isinstance(payload, dict):
            raise TransportError(f"invalid error body type: {type(payload)!r}")
        error_body = ErrorBody.from_mapping(payload)
        origin = (
            ErrorOrigin.ROUTER
            if frame.envelope.from_peer == ROUTER_PEER_ID
            else ErrorOrigin.REMOTE
        )
        return error_from_body(error_body, origin=origin)

    def _validate_target_peer(self, target: QueryGetResponseBody) -> None:
        if not target.exists or not target.connected or target.peer_class is None:
            raise RouterError(
                "peer_not_found",
                f"target peer {self._connection.target_peer_id} is not live",
            )
        if target.peer_class is not PeerClass.SERVICE:
            raise RouterError(
                "invalid_target_class",
                f"target peer {self._connection.target_peer_id} is not a service",
            )

    def _require_transport(self) -> WebSocketTransport:
        if self._transport is None or not self._transport.active:
            raise TransportError("caller handle has not been started")
        return self._transport

    async def __aenter__(self) -> Handle:
        await self.start()
        return self

    async def __aexit__(self, exc_type: Any, exc: Any, tb: Any) -> None:
        await self.stop()


def _clean_optional_str(value: Any) -> str | None:
    if value is None:
        return None
    cleaned = str(value).strip()
    return cleaned or None


def _clean_optional_int(value: Any) -> int | None:
    if value is None:
        return None
    return int(value)


def _clean_optional_float(value: Any, *, default: float) -> float:
    if value is None:
        return default
    return float(value)


from .sync import HandleSync  # noqa: E402  # isort: skip

__all__ = ["ConnectOptions", "connect", "Connection", "Handle", "HandleSync"]

File: javascript/src/connection.ts
/**
 * FILE: javascript/src/connection.ts
 * PURPOSE: Expose the caller-side v5 connection descriptor and live handle with proxy-based remote calls.
 * OWNS: Connection, Handle, caller bootstrap flow, request correlation, and query/call convenience methods.
 * EXPORTS: ConnectOptions, connect, Connection, Handle
 * DOCS: docs/spec.md, agent_chat/node_v5_api_surface_2026-03-24.md
 */

import { randomUUID } from "crypto";
import {
    decodeBody,
    encodeBody,
} from "./frame.js";
import {
    caller,
    CallBody,
    ErrorBody,
    Envelope,
    KIND_CALL,
    KIND_DISCONNECT,
    KIND_ERROR,
    KIND_QUERY,
    KIND_RESPONSE,
    QueryBody,
    QueryExistsResponseBody,
    QueryGetResponseBody,
    ResponseBody,
    service,
} from "./protocol.js";
import {
    BootstrapFailureError,
    RemoteCallError,
    RouterError,
    TransportFailureError,
} from "./errors.js";
import {
    bootstrapRouter,
    resolveRouterEndpoint,
    RouterBootstrapOptions,
} from "./router_bootstrap.js";
import {
    RoutedFrame,
    TransportCallbacks,
    TransportConnectOptions,
    WebSocketTransport,
} from "./transport.js";

const DEFAULT_REQUEST_TIMEOUT_MS = 30_000;
const DEFAULT_BOOTSTRAP_TIMEOUT_MS = 10_000;

export interface ConnectOptions {
    callerPeerId?: string;
    routerPort?: number;
    routerBin?: string;
    requestTimeoutMs?: number;
    bootstrapTimeoutMs?: number;
    eagerTargetQuery?: false | "exists" | "get";
}

type PendingDecoder<T> = (frame: RoutedFrame) => T;

interface PendingRequest<T> {
    resolve: (value: unknown) => void;
    reject: (error: Error) => void;
    decoder: PendingDecoder<T>;
    timer: ReturnType<typeof setTimeout>;
}

export function connect(targetPeerId: string, options: ConnectOptions = {}): Connection {
    return new Connection(targetPeerId, options);
}

export class Connection {
    readonly targetPeerId: string;
    readonly options: ConnectOptions;

    constructor(targetPeerId: string, options: ConnectOptions = {}) {
        this.targetPeerId = targetPeerId;
        this.options = options;
    }

    handle(): Handle {
        return new Handle(this);
    }
}

export class Handle {
    public readonly call: Record<string, (...args: unknown[]) => Promise<unknown>>;
    private readonly connection: Connection;
    private transport?: WebSocketTransport;
    private callerPeerId?: string;
    private started = false;
    private readonly pendingRequests = new Map<string, PendingRequest<unknown>>();

    constructor(connection: Connection) {
        this.connection = connection;
        this.call = new Proxy({} as Record<string, (...args: unknown[]) => Promise<unknown>>, {
            get: (_target, property) => {
                if (property === "then" || typeof property !== "string") {
                    return undefined;
                }

                return async (...args: unknown[]) => this.callRemote(property, args);
            },
        });
    }

    async start(): Promise<void> {
        if (this.started && this.transport) {
            return;
        }

        const bootstrapOptions: RouterBootstrapOptions = {
            port: this.connection.options.routerPort,
            routerBin: this.connection.options.routerBin,
            readinessTimeoutMs: this.connection.options.bootstrapTimeoutMs ?? DEFAULT_BOOTSTRAP_TIMEOUT_MS,
        };
        const bootstrap = await bootstrapRouter(bootstrapOptions);
        this.callerPeerId = this.connection.options.callerPeerId ?? `caller-${randomUUID()}`;

        const callbacks: TransportCallbacks = {
            response: frame => this.resolvePending(frame),
            error: frame => this.rejectPending(frame),
            close: error => this.rejectAllPending(
                error ?? new TransportFailureError("caller transport closed")
            ),
        };

        const transportOptions: TransportConnectOptions = {
            endpoint: bootstrap.endpoint,
            peerId: this.callerPeerId,
            peerClass: caller,
            callbacks,
            registerTimeoutMs: this.connection.options.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS,
        };

        try {
            this.transport = await WebSocketTransport.connect(transportOptions);
            this.started = true;

            if (this.connection.options.eagerTargetQuery) {
                await this.validateTargetPeer(this.connection.options.eagerTargetQuery);
            }
        } catch (error) {
            await this.stop();
            if (error instanceof Error) {
                throw error;
            }
            throw new BootstrapFailureError(String(error), { cause: error });
        }
    }

    async stop(): Promise<void> {
        const transport = this.transport;
        this.transport = undefined;
        this.started = false;

        if (!transport) {
            this.rejectAllPending(new TransportFailureError("caller handle stopped"));
            return;
        }

        try {
            if (!transport.isClosed) {
                const envelope = this.buildEnvelope(KIND_DISCONNECT, "router");
                await transport.send(envelope, new Uint8Array());
            }
        } catch {
            // Best-effort graceful disconnect only.
        } finally {
            await transport.close();
            this.rejectAllPending(new TransportFailureError("caller handle stopped"));
        }
    }

    async queryPeerExists(peerId: string): Promise<boolean> {
        const response = await this.sendRequest<QueryExistsResponseBody>(
            KIND_QUERY,
            "router",
            {
                query: "peer.exists",
                peer_id: peerId,
            } satisfies QueryBody,
            frame => decodeBody<QueryExistsResponseBody>(frame.bodyBytes)
        );
        return response.exists;
    }

    async queryPeerGet(peerId: string): Promise<QueryGetResponseBody> {
        return this.sendRequest<QueryGetResponseBody>(
            KIND_QUERY,
            "router",
            {
                query: "peer.get",
                peer_id: peerId,
            } satisfies QueryBody,
            frame => decodeBody<QueryGetResponseBody>(frame.bodyBytes)
        );
    }

    private async validateTargetPeer(mode: false | "exists" | "get"): Promise<void> {
        if (mode === "exists") {
            const exists = await this.queryPeerExists(this.connection.targetPeerId);
            if (!exists) {
                throw new RouterError(`target peer '${this.connection.targetPeerId}' does not exist`);
            }
            return;
        }

        const details = await this.queryPeerGet(this.connection.targetPeerId);
        if (!details.exists) {
            throw new RouterError(`target peer '${this.connection.targetPeerId}' does not exist`);
        }
        if (!details.connected) {
            throw new RouterError(`target peer '${this.connection.targetPeerId}' is not connected`);
        }
        if (details.class !== service) {
            throw new RouterError(`target peer '${this.connection.targetPeerId}' is not a service`);
        }
    }

    private async callRemote(functionName: string, args: unknown[]): Promise<unknown> {
        return this.sendRequest<ResponseBody>(
            KIND_CALL,
            this.connection.targetPeerId,
            {
                function: functionName,
                args,
                namespace: "default",
            } satisfies CallBody,
            frame => decodeBody<ResponseBody>(frame.bodyBytes)
        ).then(response => response.result);
    }

    private async sendRequest<T>(
        kind: typeof KIND_CALL | typeof KIND_QUERY,
        to: string,
        body: CallBody | QueryBody,
        decode: PendingDecoder<T>
    ): Promise<T> {
        if (!this.started || !this.transport || !this.callerPeerId) {
            throw new TransportFailureError("caller handle is not started");
        }

        const msgId = randomUUID();
        const envelope = this.buildEnvelope(kind, to, msgId);
        const requestTimeoutMs = this.connection.options.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS;

        return await new Promise<T>((resolve, reject) => {
            const timer = setTimeout(() => {
                this.pendingRequests.delete(msgId);
                reject(new TransportFailureError(`request timed out after ${requestTimeoutMs}ms`));
            }, requestTimeoutMs);

            this.pendingRequests.set(msgId, {
                resolve: value => resolve(value as T),
                reject,
                decoder: decode,
                timer,
            });

            void this.transport!
                .send(envelope, encodeBody(body))
                .catch(error => {
                    clearTimeout(timer);
                    this.pendingRequests.delete(msgId);
                    reject(
                        error instanceof Error
                            ? error
                            : new TransportFailureError(String(error), { cause: error })
                    );
                });
        });
    }

    private resolvePending(frame: RoutedFrame): void {
        const pending = this.pendingRequests.get(frame.envelope.msg_id);
        if (!pending) {
            return;
        }

        this.pendingRequests.delete(frame.envelope.msg_id);
        clearTimeout(pending.timer);

        try {
            pending.resolve(pending.decoder(frame));
        } catch (error) {
            pending.reject(
                error instanceof Error
                    ? error
                    : new TransportFailureError(String(error), { cause: error })
            );
        }
    }

    private rejectPending(frame: RoutedFrame): void {
        const pending = this.pendingRequests.get(frame.envelope.msg_id);
        if (!pending) {
            return;
        }

        this.pendingRequests.delete(frame.envelope.msg_id);
        clearTimeout(pending.timer);

        const errorBody = decodeBody<ErrorBody>(frame.bodyBytes);
        pending.reject(this.errorFromBody(errorBody));
    }

    private rejectAllPending(error: Error): void {
        for (const [msgId, pending] of this.pendingRequests.entries()) {
            clearTimeout(pending.timer);
            pending.reject(error);
            this.pendingRequests.delete(msgId);
        }
    }

    private errorFromBody(body: ErrorBody): Error {
        if (body.kind === "application" || body.kind === "service") {
            return new RemoteCallError(body.message, {
                details: body,
            });
        }

        return new RouterError(body.message, {
            details: body,
        });
    }

    private buildEnvelope(kind: typeof KIND_CALL | typeof KIND_QUERY | typeof KIND_DISCONNECT, to: string, msgId: string = randomUUID()): Envelope {
        if (!this.callerPeerId) {
            throw new TransportFailureError("caller identity not initialized");
        }

        return {
            v: 5,
            kind,
            msg_id: msgId,
            from: this.callerPeerId,
            to,
            ts: Date.now() / 1000,
        };
    }
}
